app/models/concerns/due_date_actions.rb

```diff
@@ -0,0 +1,300 @@
+# frozen_string_literal: true
+
+module DueDateActions
+  extend ActiveSupport::Concern
+
+  included do
+    has_many :due_dates, as: :parent, dependent: :destroy
+    include DueDateQueries
+  end
+
+  # Generic activity permission checker that determines if an activity is permissible
+  # based on the current deadline state for this parent object
+  def activity_permissible?(activity)
+    current_deadline = next_due_date
+    return false unless current_deadline
+
+    current_deadline.activity_permissible?(activity)
+  end
+
+  # Syntactic sugar methods for common activities
+  # These provide clean, readable method names while avoiding DRY violations
+  def submission_permissible?
+    activity_permissible?(:submission)
+  end
+
+  def review_permissible?
+    activity_permissible?(:review)
+  end
+
+  def teammate_review_permissible?
+    activity_permissible?(:teammate_review)
+  end
+
+  def metareview_permissible?
+    activity_permissible?(:metareview)
+  end
+
+  def quiz_permissible?
+    activity_permissible?(:quiz)
+  end
+
+  def team_formation_permissible?
+    activity_permissible?(:team_formation)
+  end
+
+  def signup_permissible?
+    activity_permissible?(:signup)
+  end
+
+  def drop_topic_permissible?
+    activity_permissible?(:drop_topic)
+  end
+
+  # Check activity permissions for a specific deadline type
+  def activity_permissible_for_type?(activity, deadline_type_name)
+    deadline = find_deadline(deadline_type_name)
+    return false unless deadline
+
+    deadline.activity_permissible?(activity)
+  end
+
+  # Get the current stage/deadline for a specific action
+  def current_stage_for(action)
+    current_deadline_for(action)
+  end
+
+  # Check if a specific action is currently allowed based on deadlines
+  def action_allowed?(action)
+    case action.to_s.downcase
+    when 'submit', 'submission'
+      submission_permissible?
+    when 'review'
+      review_permissible?
+    when 'teammate_review'
+      teammate_review_permissible?
+    when 'metareview'
+      metareview_permissible?
+    when 'quiz'
+      quiz_permissible?
+    when 'team_formation'
+      team_formation_permissible?
+    when 'signup'
+      signup_permissible?
+    when 'drop_topic'
+      drop_topic_permissible?
+    else
+      false
+    end
+  end
+
+  # Get all currently allowed actions
```

Comment:
I see no reason that this method would ever be needed.  The currently allowed actions are determined by the ..._allowed fields in the next due date.  Collecting them together serves no purpose that I can imagine.